### PR TITLE
New LKSG framework in backend

### PR DESCRIPTION
**Pull request by emanueldfine**
# Pull Request \<Title>
`<Description here>`
## Things to do during Peer Review
Please check all boxes before the Pull Request is merged. In case you skip a box, describe in the PRs description (that means: here) why the check is skipped.
- [x] The Github Actions (including Sonarqube Gateway and Lint Checks) are green. This is enforced by Github. 
- [x] A peer-review has been executed
  - [x] The code has been manually inspected by someone who did not implement the feature
- [x] The PR actually implements what is described in the JIRA-Issue
- [x] At least one E2E Test exists testing the new feature
- [x] Documentation is updated as required
- [x] The automated deployment is updated if required
- [x] The new version is deployed to the dev server using this branch
  - [x] It's verified that this version actually is the one deployed (check actuator/info for branch name and commit id!)
  - [x] The new feature is manually used/tested/observed on dev server
  - [x] All implemented Social Logins have been tested manually in the UI
- [x] If any work on the UI is to be merged, those changes were also documented in the Figma 
- [x] The local Dev stack still works: execute `startDevelopmentStack.sh`, execute npm run cypress and run Cypress Tests locally

### CODE REVIEW

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LKSGData.kt [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+/**
+ * --- API model ---
+ * Fields of the questionnaire for the LKSG framework
+ */
+@DataType("lksg")
+data class LKSGData(
+    val dataDate: LocalDate? = null,
+
+    val LKSGinScope: YesNo? = null,
```
Comment (stewal19) [inline dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LKSGData.kt:16]: When running detekt does this work? I think our detekt rules forbid the use of capital letters

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LKSGData.kt [hunk 2]
```diff
@@ -0,0 +1,221 @@
+package org.dataland.datalandbackend.model.lksg
```
Comment (stewal19) [inline dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LKSGData.kt:1]: What is this file and why is it not displayed in the IDE?

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LksgData.kt [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+    val OSHpolicyAccidentsBehaviouralResponse: YesNo? = null,
+
+    val OSHpolicyWorkplaceErgonomics: YesNo? = null,
+
+    val OSHpolicyHandlingChemicalsAndOtherHazardousSubstances: YesNo? = null,
+
+    val OSHpolicyFireProtection: YesNo? = null,
+
+    val OSHpolicyWorkingHours: YesNo? = null,
```
Comment (stewal19) [inline dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LksgData.kt:106]: detekt failing due to naming

File: dataland-frontend/tests/e2e/fixtures/lksg/LKSGdataFixtures.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
+import { faker } from "@faker-js/faker";
+import { LksgData, ProductionSite } from "../../../../build/clients/backend";
```
Comment (stewal19) [inline dataland-frontend/tests/e2e/fixtures/lksg/LKSGdataFixtures.ts:2]: Maybe use import { xyz } from @abc instead of import { xyz} from "../../../..somewhere"

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LKSGData.kt (hunk 2 above)
Comment (emanueldfine) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LKSGData.kt]: I deleted this file via GitHub commit.
I think it's not displayed in the IDE because it only exists on remote

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LKSGData.kt (hunk 1 above)
Comment (emanueldfine) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LKSGData.kt]: This file no longer exists

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LksgData.kt (hunk 3 above)
Comment (emanueldfine) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LksgData.kt:106]: Seems like Kata fixed this, but the changes got lost during merging remote on her local branch.
So waiting for re-fix.

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LKSGData.kt (hunk 2 above)
Comment (stewal19) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LKSGData.kt]: Ok

File: dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/MetaDataControllerTest.kt [hunk 5]
```diff
@@ -65,7 +65,7 @@ class MetaDataControllerTest {
         for ((company, data) in testData) {
             val testCompanyId = companyDataControllerApi.postCompany(company).companyId
             data.forEach {
-                euTaxonomyDataForNonFinancialsControllerApi.postCompanyAssociatedData(
+                euTaxonomyDataForNonFinancialsControllerApi.postCompanyAssociatedData1(
```
Comment (stewal19) [inline dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/MetaDataControllerTest.kt:68]: Would it make sense to give this another name, one that is more speaking then postCompanyAssociateData1? Like postCompanyAssociateDataTaxonomyNonFinancial or is this not possible?

File: dataland-frontend/tests/e2e/specs/prepopulation/Prepopulation.ts [hunk 6]
```diff
@@ -116,5 +118,32 @@ describe(
         checkMatchingIds(DataTypeEnum.EutaxonomyNonFinancials, companiesWithEuTaxonomyDataForNonFinancials.length);
       });
     });
+
+    describe("Upload and validate Lksg data", () => {
```
Comment (stewal19) [inline dataland-frontend/tests/e2e/specs/prepopulation/Prepopulation.ts:122]: Is it possible to unify the uploading/checking of data for all frameworks? My idea would be to have one function which hast the functionalty and then is run for every framework. This could reduce duplications in the code.

File: dataland-frontend/tests/e2e/fixtures/lksg/index.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { generateFixtureDataset } from "../FixtureUtils";
+import { generateLksgData } from "./LksgDataFixtures";
+import fs from "fs";
+import { LksgData } from "../../../../build/clients/backend";
```
Comment (stewal19) [inline dataland-frontend/tests/e2e/fixtures/lksg/index.ts:4]: See also the other comment in regards to imports

File: dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/frameworks/Lksg.kt [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        LksgDataControllerApi(BASE_PATH_TO_DATALAND_BACKEND)
+
+    private val testDataProviderForLksgData =
+        TestDataProvider(LksgData::class.java)
+
+    private val companyDataControllerApi = CompanyDataControllerApi(BASE_PATH_TO_DATALAND_BACKEND)
+
+    private val metaDataControllerApi = MetaDataControllerApi(BASE_PATH_TO_DATALAND_BACKEND)
+
+    private fun postOneCompanyAndLksg():
```
Comment (SiegfriedSobkowiak) [inline dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/frameworks/Lksg.kt:27]: This function exists 3 times. Can we please unify it to one?

File: dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+  return {
+    name: fakeSiteName,
+    isInHouseProductionOrIsContractProcessing: yesNo,
+    address: fullFormattedFakeAddress,
+    listOfGoodsAndServices: fakeGoodsAndServices,
+  };
+}
+
+export function generateArrayOfProductionSites(): ProductionSite[] {
+  return Array.from({ length: faker.datatype.number({ min: 1, max: 50 }) }, generateProductionSite);
```
Comment (SiegfriedSobkowiak) [inline dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts:33]: 50 seems to be a bit much from my point of view. Also I would include 0 since that might trigger corner cases. So maybe between 0 and 5?

File: dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    " " +
+    faker.address.buildingNumber() +
+    ", " +
+    faker.address.zipCode() +
+    " " +
+    faker.address.city() +
+    ", " +
+    faker.address.country();
+  const fakeGoodsAndServices = Array.from(
+    { length: faker.datatype.number({ min: 1, max: 5 }) },
```
Comment (SiegfriedSobkowiak) [inline dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts:20]: I would also include 0

File: dataland-frontend/tests/e2e/specs/company-metadata/SearchCompaniesForFrameworkDataDropdownFilter.ts [hunk 11]
```diff
@@ -17,7 +17,7 @@ before(function () {
 });
 
 describe("As a user, I expect the search functionality on the /companies page to adjust to the selected dropdown filters", () => {
-  it("LKSG and SFDR should be displayed in the framework dropdown even though they are not yet implemented", () => {
+  it("Lksg and Sfdr should be displayed in the framework dropdown even though they are not yet implemented", () => {
```
Comment (SiegfriedSobkowiak) [inline dataland-frontend/tests/e2e/specs/company-metadata/SearchCompaniesForFrameworkDataDropdownFilter.ts:20]: I do not get the change here. This is a string and we do not need to follow camel case convention but can use the actual names for the frameworks (LkSG and SFDR)

File: dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+  returnBase.betterWorkProgramCertificate = randomYesNoUndefined();
+  returnBase.dataDate = generateDataDate();
+  returnBase.companyLegalForm = getCompanyLegalForm();
+  returnBase.vatIdentificationNumber = generateVatIdentificationNumber();
+  returnBase.numberOfEmployees = faker.datatype.number({ min: 1000, max: 200000 });
+  returnBase.shareOfTemporaryWorkers = faker.datatype.number({ min: 0, max: 30, precision: 0.01 });
+  returnBase.totalRevenue = faker.datatype.float({ min: 10000000, max: 100000000000 });
+  returnBase.totalRevenueCurrency = generateIso4217CurrencyCode();
+  returnBase.responsibilitiesForFairWorkingConditions = randomYesNoUndefined();
```
Comment (SiegfriedSobkowiak) [inline dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts:76]: I am not a fan of this huge block of code. How about putting all the randomYesNoUndefined values into a loop. Something like for each yesNoNa value in the backend model for LkSG create an entry with the same name and a yesNoNa value in the LksgData object?

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LksgData.kt [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+    val numberOfEmployees: BigDecimal? = null,
+
+    val shareOfTemporaryWorkers: BigDecimal? = null,
+
+    val totalRevenue: BigDecimal? = null,
+
+    val totalRevenueCurrency: String? = null,
+
+    val responsibilitiesForFairWorkingConditions: YesNo? = null,
```
Comment (SiegfriedSobkowiak) [inline dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LksgData.kt:30]: In the excel specification I am counting 105 yes/no fields, in the data model here there are only 96. What is the reason for the discrepancy?

File: dataland-frontend/tests/e2e/fixtures/lksg/LKSGdataFixtures.ts (hunk 4 above)
Comment (kmachura-mika) [reply dataland-frontend/tests/e2e/fixtures/lksg/LKSGdataFixtures.ts]: Repaired.

File: dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts [hunk 14]
```diff
@@ -0,0 +1,174 @@
+import { faker } from "@faker-js/faker";
+import { LksgData, ProductionSite } from "../../../../build/clients/backend";
```
[the remark this answers is not in the document]
Comment (kmachura-mika) [reply dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts:2]: Repaired.

File: dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/MetaDataControllerTest.kt (hunk 5 above)
Comment (kmachura-mika) [reply dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/MetaDataControllerTest.kt:68]: Already changed.

File: dataland-frontend/tests/e2e/fixtures/lksg/index.ts (hunk 7 above)
Comment (kmachura-mika) [reply dataland-frontend/tests/e2e/fixtures/lksg/index.ts:4]: Repaired.

File: dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts (hunk 9 above)
Comment (kmachura-mika) [reply dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts:33]: Repaired.

File: dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts (hunk 10 above)
Comment (kmachura-mika) [reply dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts:20]: Repaired.

File: dataland-frontend/tests/e2e/specs/company-metadata/SearchCompaniesForFrameworkDataDropdownFilter.ts (hunk 11 above)
Comment (kmachura-mika) [reply dataland-frontend/tests/e2e/specs/company-metadata/SearchCompaniesForFrameworkDataDropdownFilter.ts:20]: Repaired.

File: dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LksgData.kt (hunk 13 above)
Comment (kmachura-mika) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LksgData.kt:30]: Yes, you are right. Although I think when I was working on it these fields that are missing were for uploading the report and I was supposed to omit them. Now in Excel file, there are yes or no questions. I am adding them.

Comment (kmachura-mika) [reply dataland-backend/src/main/kotlin/org/dataland/datalandbackend/model/lksg/LksgData.kt:30]: 62-70 on Jira in the file for upload and on newest Data Dictionary on Teams 62-70 now as a Yes or No question.

File: dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts (hunk 10 above)
Comment (emanueldfine) [reply dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts:20]: @SiegfriedSobkowiak but how can there be a plant, which offers 0 goods and services?
Or have I misunderstood the meaning of this field in the data dictionary?

Comment (SiegfriedSobkowiak) [reply dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts:20]: I agree there should usually be at least one thing that is being produced but I would include the 0 to test unexpected input.

File: dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts (hunk 12 above)
Comment (SiegfriedSobkowiak) [reply dataland-frontend/tests/e2e/fixtures/lksg/LksgDataFixtures.ts:75]: Emanuel and I had a look into this and it turns out it is not possible to do this using typescript

File: dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/frameworks/Lksg.kt [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+        Pair<DataMetaInformation, LksgData> {
+        tokenHandler.obtainTokenForUserType(TokenHandler.UserType.Admin)
```
[the remark this answers is not in the document]
Comment (kmachura-mika) [reply dataland-e2etests/src/test/kotlin/org/dataland/e2etests/tests/frameworks/Lksg.kt:29]: Repaired.

File: dataland-frontend/tests/e2e/specs/prepopulation/Prepopulation.ts (hunk 6 above)
Comment (SiegfriedSobkowiak) [reply dataland-frontend/tests/e2e/specs/prepopulation/Prepopulation.ts:128]: We looked into it and the tests are already generic for most of the functionality, the only part that could be improved in my opinion is the loading of the fixture data. I will look into it.

Comment (SiegfriedSobkowiak) [reply dataland-frontend/tests/e2e/specs/prepopulation/Prepopulation.ts:128]: It does not add really to the code quality and hece will be skipped.
